cilium/CFP-32759-cilium-mtls-via-mtls-agent.md

```diff
@@ -0,0 +1,304 @@
+
+Other controllers or control planes could be responsible for providing configuration to the mTLS agent. However, utilizing the existing Cilium agent for configuration minimizes the addition of new components and simplifies design. The Cilium agent has an existing xDS server for providing configuration to Envoy which could be generalized to support xDS requests from the mTLS agent. Alternatively, a new xDS server implementation specialized for this feature could be added.
+
+### Datapath 
+
+We describe an example of the datapath involved in this feature proposal. **The mTLS agent is in its own network namespace and the eBPF component is performing L2 redirection**. For brevity, we will trace a single packet on a typical cross-node communication pathway. Consider the following cluster:
+* There are two nodes, Node A and Node B.
+* Pod A, mTLS Agent A and Cilium Agent A are on Node A.
+* Pod B, mTLS Agent B and Cilium Agent B are on Node B.
+
+The packet is being sent from Pod A to Pod B, both of which are mTLS-enabled.
+
+![alt text](images/cilium-mTLS-cluster-diagram.png)
+
+#### Step 1: From Pod A to mTLS Agent A
+
+A request leaves Pod A, the mTLS-enabled client, going from its eth0 interface to the node network namespace via the veth device. At this veth, the egress from-container BPF program will be invoked, which will result in a tail call to a function for packet processing.
+
+When the packet is processed, the BPF program will check the node-local BPF map for redirection. Recall, this map’s entries are of the form {podIP : (interface index, interface MAC addr)} and will contain a map entry for every mTLS-enabled pod on the node. Since Pod A is mTLS-enabled, its IP will be present as a key. This indicates that we should look up the mTLS Agent in the map using its special well-known key. Indexing via a well-known key allows the BPF program to be unaware of the IP address of the mTLS Agent. The packet’s destination MAC address is then rewritten to mTLS Agent A’s MAC address, and the packet is redirected to the appropriate network interface.
+
+![alt text](images/cilium-mTLS-bpf-redirection-map.png)
+
+The packet will reach the mTLS agent’s pod network namespace via a veth pair.. In the mTLS agent’s network namespace of this POC, iptables TPROXY rules have been added to the mangle table’s PREROUTING chain. A [TPROXY](https://www.kernel.org/doc/html/latest/networking/tproxy.html) rule here will deliver the packet to the destination address 127.0.0.1:15001 and mark the packet. The marked packet will use a special routing table, which contains a rule that will route it to the local loopback device. The packet will thereby reach the process listening on the mTLS agent at port 15001. 
```

Comment:
why it requires a special routing table? when tproxy redirects to localhost:15001, wont the packet is delivered to mtls agent?